src/Lykke.AzureStorage/Tables/AzureTableStorage.cs

```diff
@@ -435,6 +435,46 @@ IEnumerator IEnumerable.GetEnumerator()
             }
         }
 
+        /// <inheritdoc/>
+        public async Task<bool> InsertOrReplaceAsync(T entity, Func<T, bool> replaceCondition)
+        {
+            while (true)
+            {
+                try
+                {
+                    var table = await GetTableAsync();
+                    var existingEntity = await GetDataAsync(entity.PartitionKey, entity.RowKey);
+
+                    if (existingEntity != null)
+                    {
+                        if (!replaceCondition(existingEntity))
+                        {
+                            return false;
+                        }
+
+                        entity.ETag = existingEntity.ETag;
+
+                        await table.ExecuteAsync(TableOperation.Replace(entity), GetRequestOptions(), null);
+                    }
+                    else
+                    {
+                        await table.ExecuteAsync(TableOperation.Insert(entity), GetRequestOptions(), null);
+                    }
+
+                    return true;
+                }
+                catch (StorageException e) when (e.RequestInformation.HttpStatusCode == 412)
+                {
+                }
+                catch (StorageException e) when (e.RequestInformation.HttpStatusCode == 409)
+                {
+                }
+                catch (StorageException e) when (e.RequestInformation.HttpStatusCode == 404)
```

Comment:
404 - from this - https://docs.microsoft.com/en-us/rest/api/storageservices/table-service-error-codes - 404 is returned only for 'Table is not found' case. May be we need to throw exception in this case, right?

Reply:
good comment, need check other methods for this

Reply:
Yes, but here - https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes there is ResourceNotFound. Replace can responce 404, if row has been deleted by another thread.